server/src/main/java/org/eclipse/lsp/cobol/service/CFASTBuilderImpl.java

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright (c) 2021 Broadcom.
+ * The term "Broadcom" refers to Broadcom Inc. and/or its subsidiaries.
+ *
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *    Broadcom, Inc. - initial API and implementation
+ *
+ */
+package org.eclipse.lsp.cobol.service;
+
+import org.eclipse.lsp.cobol.core.CobolParser;
+import org.eclipse.lsp.cobol.core.model.extendedapi.*;
+import org.eclipse.lsp.cobol.core.model.tree.*;
+import org.eclipse.lsp4j.Position;
+import org.eclipse.lsp4j.Range;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * CF tree builder iplementation
+ */
+public class CFASTBuilderImpl implements CFASTBuilder {
+  private static final int SNIPPET_LENGTH = 10;
+  //  IF statement without END_IF workaround. Should be removed after grammar fixed.
+  private boolean needCloseIf = false;
+
+  @Override
+  public ExtendedApiResult build(Node rootNode) {
+    ExtendedApiResult result = new ExtendedApiResult(new ArrayList<>());
+    if (rootNode == null) {
+      return result;
+    }
+    for (Node node : rootNode.getChildren()) {
+      traverse((ProgramNode) node, result.getControlFlowAST());
+    }
+    return result;
+  }
+
+  private void traverse(CFASTNode parent, Node node) {
+    if (node instanceof ParagraphNode) {
```

Comment:
Well...
Without `else` it would have to enter every condition here.

Reply:
Actually, it's not. Compile will optimize it. But ok, I'll fix it.

Reply:
Makes sense. My main doubt here is that I don't like conditions based on types because it is not very effective and lacks maintainability. May this logic be encapsulated in the classes?

Reply:
ParagraphNode which could contain this logic is a syntax tree node class, and in this case, CF AST logic will leak into Syntax Tree which shouldn't depend on its users.
What if I'll split it into methods in CFASTBuilderImpl?
It will not look nice for ```node instanceof ExitNode || node instanceof GoBackNode || node instanceof StopNode``` though

Reply:
Ok, I agree